File: tests/lib/git.test.js
const { parseRepo, buildCloneUrl, getRepoInfo } = require('../../lib/git');
const fs = require('fs');
const path = require('path');
const os = require('os');

describe('Git Module', () => {
  describe('parseRepo', () => {
    test('should parse valid owner/repo format', () => {
      const result = parseRepo('facebook/react');
      expect(result).toEqual({
        owner: 'facebook',
        name: 'react'
      });
    });

    test('should throw error for invalid format - missing owner', () => {
      expect(() => parseRepo('react')).toThrow('Invalid repo format');
    });

    test('should throw error for invalid format - missing name', () => {
      expect(() => parseRepo('facebook/')).toThrow('Invalid repo format');
    });

    test('should throw error for empty string', () => {
      expect(() => parseRepo('')).toThrow('Invalid repo format');
    });

    test('should handle organizations with hyphens', () => {
      const result = parseRepo('vercel/next.js');
      expect(result).toEqual({
        owner: 'vercel',
        name: 'next.js'
      });
    });
  });

  describe('buildCloneUrl', () => {
    test('should build correct HTTPS URL', () => {
      const url = buildCloneUrl('facebook/react');
      expect(url).toBe('https://github.com/facebook/react.git');
    });

    test('should build URL for organization with hyphens', () => {
      const url = buildCloneUrl('some-org/project-name');
      expect(url).toBe('https://github.com/some-org/project-name.git');
    });

    test('should handle proxy parameter (for future use)', () => {
      const url = buildCloneUrl('facebook/react', 'http://proxy:8080');
      expect(url).toBe('https://github.com/facebook/react.git');
    });
  });

  describe('getRepoInfo', () => {
    let tempDir;

    beforeEach(() => {
      // Create temporary directory
      tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'gh-test-'));
    });

    afterEach(() => {
      // Cleanup temp directory
      if (fs.existsSync(tempDir)) {
        fs.rmSync(tempDir, { recursive: true, force: true });
      }
    });

    test('should return null for non-existent directory', () => {
      const result = getRepoInfo('/non/existent/path');
      expect(result).toBeNull();
    });

    test('should return null for directory without .git', () => {
      const result = getRepoInfo(tempDir);
      expect(result).toBeNull();
    });
  });
});

File: internal/git/git_test.go
package git_test

import (
	"testing"

	"github.com/syxc/ghr/internal/git"
)

func TestParseRepo(t *testing.T) {
	tests := []struct {
		input    string
		wantOwn  string
		wantName string
		wantErr  bool
	}{
		{"facebook/react", "facebook", "react", false},
		{"vercel/next.js", "vercel", "next.js", false},
		{"some-org/project-name", "some-org", "project-name", false},
		{"react", "", "", true},
		{"facebook/", "", "", true},
		{"", "", "", true},
		{"/repo", "", "", true},
	}

	for _, tt := range tests {
		t.Run(tt.input, func(t *testing.T) {
			owner, name, err := git.ParseRepo(tt.input)
			if tt.wantErr {
				if err == nil {
					t.Errorf("ParseRepo(%q) expected error, got nil", tt.input)
				}
				return
			}
			if err != nil {
				t.Errorf("ParseRepo(%q) unexpected error: %v", tt.input, err)
				return
			}
			if owner != tt.wantOwn || name != tt.wantName {
				t.Errorf("ParseRepo(%q) = (%q, %q), want (%q, %q)", tt.input, owner, name, tt.wantOwn, tt.wantName)
			}
		})
	}
}

func TestBuildCloneURL(t *testing.T) {
	tests := []struct {
		owner string
		name  string
		want  string
	}{
		{"facebook", "react", "https://github.com/facebook/react.git"},
		{"some-org", "project-name", "https://github.com/some-org/project-name.git"},
	}

	for _, tt := range tests {
		t.Run(tt.owner+"/"+tt.name, func(t *testing.T) {
			got := git.BuildCloneURL(tt.owner, tt.name)
			if got != tt.want {
				t.Errorf("BuildCloneURL(%q, %q) = %q, want %q", tt.owner, tt.name, got, tt.want)
			}
		})
	}
}

func TestGetRepoInfo_NonExistent(t *testing.T) {
	info := git.GetRepoInfo("/non/existent/path")
	if info != nil {
		t.Error("GetRepoInfo for non-existent path should return nil")
	}
}

func TestGetRepoInfo_NoGitDir(t *testing.T) {
	info := git.GetRepoInfo(t.TempDir())
	if info != nil {
		t.Error("GetRepoInfo for non-git directory should return nil")
	}
}
